presto-druid/src/main/java/com/facebook/presto/druid/segment/ZipIndexFileSource.java

```diff
@@ -0,0 +1,281 @@
+    public static final int LOCAL_HEADER_EXTRA_FIELD_LENGTH_OFFSET = 28;
+
+    private final DataInputSource dataInputSource;
+
+    private ZipFileData zipData;
+
+    public ZipIndexFileSource(DataInputSource dataInputSource)
+    {
+        this.dataInputSource = requireNonNull(dataInputSource, "dataInputSource is null");
+        zipData = readCentralDirectory();
+    }
+
+    /**
+     * Reads file inside a zip archive
+     */
+    @Override
+    public byte[] readFile(String fileName)
+            throws IOException
+    {
+        ZipFileEntry entry = zipData.getEntry(fileName);
+        if (entry == null) {
+            throw new PrestoException(DRUID_SEGMENT_LOAD_ERROR, format("Zip doesn't contain file: %s", fileName));
+        }
+        byte[] fileData = new byte[(int) entry.getSize()];
+        readFully(entry, 0, fileData, 0, fileData.length);
+        return fileData;
+    }
+
+    @Override
+    public final void readFile(String fileName, long position, byte[] buffer)
+            throws IOException
+    {
+        ZipFileEntry entry = zipData.getEntry(fileName);
+        if (entry == null) {
+            throw new PrestoException(DRUID_SEGMENT_LOAD_ERROR, format("Zip doesn't contain file: %s", fileName));
+        }
+        readFully(entry, position, buffer, 0, buffer.length);
+    }
+
+    private void readFully(ZipFileEntry entry, long position, byte[] buffer, int bufferOffset, int bufferLength)
+            throws IOException
+    {
+        long offset = entry.getLocalHeaderOffset();
+        byte[] fileHeader = new byte[LOCAL_HEADER_FIXED_DATA_SIZE];
+        dataInputSource.readFully(offset, fileHeader);
+        offset += fileHeader.length;
+
+        if (!ZipUtil.arrayStartsWith(fileHeader, ZipUtil.intToLittleEndian(LOCAL_HEADER_SIGNATURE))) {
+            throw new PrestoException(
+                    DRUID_SEGMENT_LOAD_ERROR,
+                    format("The file '%s' is not a correctly formatted zip file: Expected a File Header at offset %d, but not present.", entry.getName(), offset));
+        }
+
+        // skip name and extra field
+        int nameLength = ZipUtil.getUnsignedShort(fileHeader, LOCAL_HEADER_FILENAME_LENGTH_OFFSET);
+        int extraFieldLength = ZipUtil.getUnsignedShort(fileHeader, LOCAL_HEADER_EXTRA_FIELD_LENGTH_OFFSET);
+        offset += (nameLength + extraFieldLength);
+
+        // deflate
+        int compressedSize = (int) entry.getCompressedSize();
+        byte[] compressedData = new byte[compressedSize];
+        dataInputSource.readFully(offset, compressedData);
+        InflaterInputStream inflaterInputStream = new InflaterInputStream(new ByteArrayInputStream(compressedData), new Inflater(true));
+
+        try {
+            inflaterInputStream.skip(position);
+            int size = chunkedRead(inflaterInputStream, buffer, bufferOffset, bufferLength);
+            checkState(size == bufferLength);
+        }
+        catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)